actix-http/src/lib.rs

```diff
@@ -81,3 +80,36 @@ pub enum Protocol {
 }
 
 type ConnectCallback<IO> = dyn Fn(&IO, &mut Extensions);
+
+// container for data that extract with ConnectCallback
+pub(crate) struct OnConnectData(Option<Extensions>);
+
+impl Default for OnConnectData {
+    fn default() -> Self {
+        Self(None)
+    }
+}
+
+impl OnConnectData {
+    // construct self from io.
+    pub(crate) fn from_io<T>(
+        io: &T,
+        on_connect_ext: Option<&ConnectCallback<T>>,
+    ) -> Self {
+        let ext = on_connect_ext.map(|handler| {
+            let mut extensions = Extensions::new();
+            handler(io, &mut extensions);
+            extensions
+        });
+
+        Self(ext)
+    }
+
+    // merge self to given request's head extension.
+    #[inline]
+    pub(crate) fn merge(&mut self, req: &mut Request) {
```

Comment:
just to make the merge direction clear when reading the dispatcher: `merge_into`, `drain_into`, `drain_into_request`, etc.

```suggestion
    pub(crate) fn merge_into(&mut self, req: &mut Request) {
```